### ISSUE TITLE
Spelunky Classic HD, GM Loader - After being remapped, the controls will not work the next time you start the game.

### ISSUE THREAD

**Issue by vic-viper-001** [Exarkuniv/RetroPie-Extra#106]
When you install gmloader, two games are also installed. Maldita Castilla and Spelunky Classic HD.

I didn't like the default button mapping for Spelunky HD Classic, so I changed it. It works great WHILE you are playing the game... exactly like it should. 

A "gamepad.json" file is generated upon doing this, and contains the following text:
{ "joyPayVal": 32773, "joyJumpVal": 32769, "joyRunVal": 32774, "joyRopeVal": 32772, "joyAttackVal": 32771, "joyBombVal": 32770, "joyStartVal": 32778, "joyFlareVal": 32772, "joyItemVal": 32777 }

If I quit the game and come back to it later, only the D-Pad will work. None of the action buttons work, and I'm forced to use a keyboard to navigate the menus to either remap the buttons yet again, or quit. For reference I use an 8bitdo sn30pro+ controller in Pro-controller mode.

is there any way to force Spelunky to "remember' my remap setting?

**Comment by Exarkuniv** [comment]
@s1eve-mcdichae1 would be the one to take a look at this, if there is any thing we can do or not

**Comment by s1eve-mcdichae1** [comment]
I don't think there's much we can do about this:

https://github.com/yancharkin/SpelunkyClassicHD/issues/19

Fortunately, however, **gmloader** uses SDL_GameController input, so while there might not be a proper "fix," at least there's a hack/workaround possible: you can remap the buttons by exporting an environment variable. You need SDL's `controllermap` utility:

```
git clone https://github.com/retropie/sdl 
cd sdl 
git checkout retropie-2.0.10
cd test
./configure && make controllermap
```

(Optional) retain the following files to a permanent location:

    controllermap
    controllermap.bmp
    axis.bmp
    button.bmp

Run `./controllermap` to list all connected joypads. Run `./controllermap {n}` to map joypad number **{n}**. Press each button/axis when prompted (use a keyboard to skip inputs that aren't on your controller with SPACE or to correct a mistake with BACKSPACE.) At the end it will output a string; you can copy this string in your SSH client, or run `./controllermap {n} > MyCustomGamepad.txt` to save output to a file called `MyCustomGamepad.txt` (or use whatever other name you want.)

Export that string as an environment variable. You can paste in the whole string, or with command substitution:

    export SDL_GAMECONTROLLERCONFIG="$(cat "/path/to/MyCustomGamepad.txt")"

You can put this line in your `autostart.sh` (before `emulationstation #auto`), for example to map a generic gamepad that is not recognized correctly by SDL.

To remap a specific game, copy the file to, for example `SpelunkyRemap.txt` and swap the B and X values (or however you prefer -- don't forget it's Xbox layout, not Nintendo, for ABXY). Edit your `roms/ports/Spelunky Classic HD.sh` launch script and export the variable in the same manner:

```
#!/bin/bash

export SDL_GAMECONTROLLERCONFIG="$(cat "/path/to/SpelunkyRemap.txt")"

"/opt/retropie/supplementary/runcommand/runcommand.sh" 0 _PORT_ "droidports" "/home/pi/RetroPie/roms/ports/droidports/spelunky_classic_hd-android.apk"
```

This modified variable exists only in the particular shell which called it and any shells that are children of that shell, so when you quit the game and return to EmulationStation, it will revert to the previously-set value (if any, or to SDL defaults) for any other games.

In-game, it still *says* "press B to use your whip," but it works when I press "X" instead. This might be the best we can do under circumstances.

**Comment by Exarkuniv** [comment]
Ok so this way is almost the same as xboxdrv. 
Cool that theres a few different options to do remapping.

**Comment by s1eve-mcdichae1** [comment]
Yeah it's pretty similar. **xboxdrv** is universal and more versatile (can map mouse and keyboard actions to a gamepad, for example, while this can only remap buttons on the gamepad); on the other hand this environment variable only works for ports that use SDL_GameController input, but you don't have to install anything extra (I mean besides the mapping utility to generate the string, I guess. But as far as actually *implementing* it, you don't have to install or "run" anything, just export a variable with the shell builtin, and SDL takes care of the rest.)

**Comment by vic-viper-001** [comment]
Thanks for the detailed instructions. I tried to follow them, but eventually reached a point where my pi went completely berserk with an error message  "permission denied, failed to execute dumb buffer" repeating endlessly, until I pressed ctrl-c. I tried it a few more times afterward, and had the same result each time. I doubt I'll ever get this to work, and have instead looked into using the 8bido ultimate software to see if I can remap my controller that way instead
